src/main/java/org/folio/search/service/systemuser/SystemUserService.java

```diff
@@ -0,0 +1,186 @@
+import static org.folio.search.client.PermissionsClient.Permissions;
+import static org.folio.search.client.UsersClient.User;
+import static org.folio.spring.scope.FolioExecutionScopeExecutionContextManager.beginFolioExecutionContext;
+import static org.folio.spring.scope.FolioExecutionScopeExecutionContextManager.endFolioExecutionContext;
+import static org.springframework.util.CollectionUtils.isEmpty;
+
+import java.nio.file.Files;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Supplier;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+import org.folio.search.client.AuthnClient;
+import org.folio.search.client.PermissionsClient;
+import org.folio.search.client.UsersClient;
+import org.folio.search.configuration.properties.FolioSystemUserProperties;
+import org.folio.search.model.SystemUser;
+import org.folio.search.repository.SystemUserRepository;
+import org.folio.search.repository.SystemUserTokenCache;
+import org.folio.search.service.context.FolioExecutionContextBuilder;
+import org.folio.spring.FolioExecutionContext;
+import org.folio.spring.integration.XOkapiHeaders;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.ResourceUtils;
+
+@Log4j2
+@RequiredArgsConstructor
+@Service
+@EnableConfigurationProperties(FolioSystemUserProperties.class)
+public class SystemUserService {
+  private final PermissionsClient permissionsClient;
+  private final UsersClient usersClient;
+  private final AuthnClient authnClient;
+  private final SystemUserRepository systemUserRepository;
+  private final FolioExecutionContextBuilder contextBuilder;
+  private final FolioSystemUserProperties folioSystemUserConf;
+  private final SystemUserTokenCache tokenCache;
+
+  public void prepareSystemUser(FolioExecutionContext context) {
+    var folioUser = getFolioUser(folioSystemUserConf.getUsername());
+    var userId = folioUser.map(User::getId)
+      .orElse(UUID.randomUUID().toString());
+
+    if (folioUser.isPresent()) {
+      addPermissions(userId);
+    } else {
+      createFolioUser(userId);
+      saveCredentials();
+      assignPermissions(userId);
+    }
+
+    systemUserRepository.save(SystemUser.builder()
+      .id(userId)
+      .username(folioSystemUserConf.getUsername())
+      .tenantId(context.getTenantId())
+      .okapiUrl(context.getOkapiUrl())
+      .build());
+  }
+
+  public SystemUser getSystemUser(String tenantId) {
+    var systemUser = findSystemUser(tenantId)
+      .orElseThrow(() -> new IllegalStateException("There is no system user configured"));
+
+    if (tokenCache.hasTokenForTenant(tenantId)) {
+      return tokenCache.getByTenant(tenantId);
+    }
+
+    return issueTokenForSystemUser(systemUser);
+  }
```

Comment:
If `tokenCache` has the token, I do not think you need `findSystemUser` first

Reply:
Thanks, you're right.